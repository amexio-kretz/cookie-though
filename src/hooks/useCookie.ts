import { CookieOption } from '../types';
import { CookiePreferences } from '../components/app';

export const COOKIE_PREFERENCES_KEY = 'cookie-preferences';

interface Options {
  cookieOptions: CookieOption[];
  cookiePreferenceKey?: string;
}

export const getCookie = (cookieKey: string): CookiePreferences | undefined => {
  const rawCookies = decodeURIComponent(document.cookie).split(';');
  return rawCookies.reduce((cookiePreferences, cookie) => {
    if (cookiePreferences) return cookiePreferences;

    const [key, value] = cookie.split('=');
    if (key === cookieKey) return JSON.parse(value);
  }, undefined);
};

export const getNextYear = () => {
  const now = new Date();
  return new Date(now.setFullYear(now.getFullYear() + 1));
};

const policiesHaveChanged = (
  policyIds: CookieOption['id'][],
  preferenceIds: CookieOption['id'][],
) => {
  return policyIds.some(id => !policyIds.includes(id)) || policyIds.length !== preferenceIds.length;
};

const mergePolicies = (policies: CookieOption[], preferences: CookieOption[]): CookieOption[] => {
  return policies.map(policy => {
    const isEnabled = preferences.reduce((isPreferenceEnabled, preference) => {
      return preference.id === policy.id ? preference.isEnabled : isPreferenceEnabled;
    }, false);

    return { ...policy, isEnabled };
  });
};

const useCookie = ({ cookieOptions, cookiePreferenceKey = COOKIE_PREFERENCES_KEY }: Options) => {
  const preferences: CookiePreferences = {
    cookieOptions,
    isCustomised: false,
  };

  const setCookiePreferences = (cookiePreferences: CookiePreferences) => {
    const expires = getNextYear();
    document.cookie = `${cookiePreferenceKey}=${JSON.stringify(
      cookiePreferences,
    )}; expires=${expires}; path=/`;

    return cookiePreferences;
  };

  const getCookiePreferences = () => {
    const cookiePreferences = getCookie(cookiePreferenceKey);

    if (!cookiePreferences) return setCookiePreferences(preferences);

    const policyIds = preferences.cookieOptions.map(cookieOption => cookieOption.id);
    const preferenceIds = cookiePreferences.cookieOptions.map(cookieOption => cookieOption.id);
    if (!policiesHaveChanged(policyIds, preferenceIds)) return cookiePreferences;

    const cookieOptions = mergePolicies(preferences.cookieOptions, cookiePreferences.cookieOptions);

    return { cookieOptions, isCustomised: false };
  };

  return { getCookiePreferences, setCookiePreferences };
};

export default useCookie;
