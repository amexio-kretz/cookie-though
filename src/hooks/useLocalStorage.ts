import { CookieOption } from '../types';
import { CookiePreferences } from './../components/app';

export const COOKIE_PREFERENCES_KEY = 'cookie-preferences';

const useLocalStorage = (cookieOptions: CookieOption[]) => {
  const DEFAULT_COOKIE_PREFERENCES: CookiePreferences = {
    cookieOptions,
    isCustomised: false,
  };

  const setCookiePreferences = (cookiePreferences: CookiePreferences) => {
    localStorage.setItem(COOKIE_PREFERENCES_KEY, JSON.stringify(cookiePreferences));
    return cookiePreferences;
  };

  const getCookiePreferences = () => {
    const rawCookiePreferences = localStorage.getItem(COOKIE_PREFERENCES_KEY);
    return rawCookiePreferences
      ? (JSON.parse(rawCookiePreferences) as CookiePreferences)
      : setCookiePreferences(DEFAULT_COOKIE_PREFERENCES);
  };

  return { getCookiePreferences, setCookiePreferences };
};

export default useLocalStorage;
