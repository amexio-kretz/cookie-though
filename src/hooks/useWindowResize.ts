import { useLayoutEffect } from 'preact/hooks';
import throttle from 'lodash.throttle';

const useWindowResize = (callback: () => void) => {
  useLayoutEffect(() => {
    const throttledCallback = throttle(callback, 100);
    window.addEventListener('resize', throttledCallback);

    return () => window.removeEventListener('resize', throttledCallback);
  }, [callback]);
};

export default useWindowResize;
