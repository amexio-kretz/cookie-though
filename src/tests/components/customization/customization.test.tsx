import { h } from 'preact';
import toJson from 'enzyme-to-json';
import { mount, ReactWrapper, shallow } from 'enzyme';
import { Collapse } from 'react-collapse';
import mockCookies from '../../__mocks__/cookieOptions';

import Button from '../../../components/button';
import Customization from '../../../components/customization';
import Slider from '../../../components/customization/slider';
import ToggleButton from '../../../components/customization/toggleButton';
import { COOKIE_PREFERENCES_KEY } from '../../../hooks/useLocalStorage';

const defaultProps = {
  cookieOptions: mockCookies,
};

describe('Customization', () => {
  it('should render properly', () => {
    const wrapper = shallow(<Customization {...defaultProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  afterEach(() => {
    defaultProps.cookieOptions = mockCookies.map(mockCookie => ({
      ...mockCookie,
      isEnabled: false,
    }));
  });

  const getToggleButton = (wrapper: ReactWrapper) => wrapper.find(ToggleButton).find('button');

  const isCustomizationCollapsed = (wrapper: ReactWrapper) => {
    return !wrapper.find(Collapse).prop('isOpened');
  };

  const getOptionalCookie = (wrapper: ReactWrapper) => wrapper.find('.option').at(1);

  const getCookiePreferences = () => {
    const cookiePreferences = localStorage.getItem(COOKIE_PREFERENCES_KEY);
    return cookiePreferences ? JSON.parse(cookiePreferences) : 'NO_PREFERENCES';
  };

  test('clicking the ToggleButton component toggles the isActive state', () => {
    const wrapper = mount(<Customization {...defaultProps} />);
    const acceptButtonLabel = () => {
      return wrapper
        .find(Button)
        .last()
        .find('button')
        .text();
    };

    getToggleButton(wrapper)
      .simulate('click')
      .update();
    expect(getToggleButton(wrapper).hasClass('active')).toBeTruthy();
    expect(isCustomizationCollapsed(wrapper)).toBeFalsy();
    expect(acceptButtonLabel()).toEqual('Accept');

    getToggleButton(wrapper)
      .simulate('click')
      .update();
    expect(getToggleButton(wrapper).hasClass('active')).toBeFalsy();
    expect(isCustomizationCollapsed(wrapper)).toBeTruthy();
    expect(acceptButtonLabel()).toEqual('Accept all');
  });

  test('clicking the Slider in an Option will toggle the isEnabled state for that option', () => {
    const wrapper = mount(<Customization {...defaultProps} />);
    const getSlider = () => {
      return getOptionalCookie(wrapper)
        .find(Slider)
        .find('button');
    };

    getSlider()
      .simulate('click')
      .update();
    expect(getOptionalCookie(wrapper).hasClass('enabled')).toBeTruthy();

    getSlider()
      .simulate('click')
      .update(),
      expect(getOptionalCookie(wrapper).hasClass('enabled')).toBeFalsy();
  });

  test('clicking the decline Button will decline all options and save the preferences', () => {
    const enabledCookies = mockCookies.map(mockCookie => ({ ...mockCookie, isEnabled: true }));
    const wrapper = mount(<Customization cookieOptions={enabledCookies} />);
    const declineButton = wrapper.find('button.secondary');

    declineButton.simulate('click').update();
    expect(getCookiePreferences()).toEqual({ isCustomised: true, cookieOptions: mockCookies });
    expect(getToggleButton(wrapper).hasClass('active')).toBeFalsy();
    expect(isCustomizationCollapsed(wrapper)).toBeTruthy();
    const optionalCookies = wrapper.find('.option').not('.optionSecondary');
    optionalCookies.forEach(optionalCookie => {
      expect(optionalCookie.hasClass('enabled')).toBeFalsy();
    });
  });

  describe('when clicking the accept button', () => {
    it('will only accept the options a user has enabled and save the preferences', () => {
      const customisedCookies = [
        { ...mockCookies[0], isEnabled: true },
        ...mockCookies.splice(1, mockCookies.length),
      ];
      const wrapper = mount(<Customization cookieOptions={customisedCookies} />);
      const acceptButton = wrapper.find('button').last();

      acceptButton.simulate('click').update();
      expect(getCookiePreferences()).toEqual({
        isCustomised: true,
        cookieOptions: customisedCookies,
      });
      expect(getToggleButton(wrapper).hasClass('active')).toBeFalsy();
      expect(isCustomizationCollapsed(wrapper)).toBeTruthy();
      expect(getOptionalCookie(wrapper).hasClass('enabled')).toBeTruthy();
    });

    it('will accept all options if the user has none enabled and save the preferences', () => {
      const wrapper = mount(<Customization {...defaultProps} />);
      const acceptButton = wrapper.find('button').last();

      acceptButton.simulate('click').update();
      expect(getCookiePreferences()).toEqual({
        isCustomised: true,
        cookieOptions: mockCookies.map(mockCookie => ({ ...mockCookie, isEnabled: true })),
      });
      expect(getToggleButton(wrapper).hasClass('active')).toBeFalsy();
      expect(isCustomizationCollapsed(wrapper)).toBeTruthy();
      const optionalCookies = wrapper.find('.option').not('.optionSecondary');
      optionalCookies.forEach(optionalCookie => {
        expect(optionalCookie.hasClass('enabled')).toBeTruthy();
      });
    });
  });
});
