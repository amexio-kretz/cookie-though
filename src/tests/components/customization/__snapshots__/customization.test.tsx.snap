// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customization should render properly 1`] = `
<div>
  <ToggleButton
    isActive={false}
    label="Customize"
    toggleCustomization={[Function]}
  />
  <Collapse
    isOpen={false}
    rootStyles={
      Object {
        "bottom": 1,
        "height": 1,
      }
    }
  >
    <Options
      isOpen={false}
      onToggle={[Function]}
      options={
        Array [
          Object {
            "category": "essential",
            "description": "Always on",
            "id": "essential",
            "isEnabled": false,
            "label": "Strictly necessary cookies",
          },
          Object {
            "category": "functional",
            "description": "We’ll remember the basics such as language.",
            "id": "functional",
            "isEnabled": false,
            "label": "Functional cookies",
          },
          Object {
            "category": "statistics",
            "description": "We’ll know where we should improve your experience.",
            "id": "statistics",
            "isEnabled": false,
            "label": "Statistics",
          },
          Object {
            "category": "marketing",
            "description": "We’ll only show you content that interests you.",
            "id": "marketing",
            "isEnabled": false,
            "label": "Marketing",
          },
        ]
      }
    />
  </Collapse>
  <div
    className="ct-acceptance"
  >
    <Button
      className="ct-button-secondary"
      label="Decline"
      onClick={[Function]}
    />
    <Button
      label="Accept all"
      onClick={[Function]}
    />
  </div>
</div>
`;
